### PR TITLE
Fixes framebuffer caching with multiple references of different sizes

### PR DESCRIPTION
**Pull request by arodic**
#Change summary

## The Root Cause of #32689

* `ViewportDepthTextureNode`: Uses a singleton _sharedDepthbuffer that is shared across all RenderTargets/CanvasTargets. This causes brute-force destruction and re-creation of depth buffer per reference on every rendered frame if references are not exactly the same size. This affects both CanvasTarget and RenderTarget

* `ViewporTextureNode` Does not consider CanvasTarget as a reference. It simply does not do caching for CanvasTargets

* `Renderer`: Uses a single _frameBufferTarget that resizes for each canvas (if not exactly the same size), triggering
depthTexture.needsUpdate = true and causing texture recreation

The Fix for #32689

* `ViewportDepthTextureNode` removed getTextureForReference() override to allow texture caching. Now it caches depth textures per reference

* `ViewportTextureNode` Added check for CanvasTarget. Previous implementation considered RenderTargets only.

* `Renderer` Changed _frameBufferTarget to _frameBufferTargets (Map<CanvasTarget, RenderTarget>). Each canvas target now gets its own framebuffer target instead of sharing one

This allows multi-canvas rendering with different sizes and tone mapping enabled without constant texture buffer recreation.

### CODE REVIEW

File: src/nodes/display/ViewportTextureNode.js [hunk 1]
```diff
@@ -130,6 +121,15 @@ class ViewportTextureNode extends TextureNode {
 
 		}
 
+		let cacheTextures = _textureCaches.get( defaultFramebuffer );
+
+		if ( cacheTextures === undefined ) {
+
+			cacheTextures = new WeakMap();
```
Comment (arodic) [inline src/nodes/display/ViewportTextureNode.js]: Looking at this WeakMap() I just realized that `ViewportTextureNode` doesn't really have a way to free up GPU resources. WeakMap afaik does not free up texture memory. This is not a new issue, previous version also had a WeakMap and lacked disposal mechanism just something to keep an eye on.

**Comment by arodic** [comment]
I cleaned up all the slop. This should be minimal and pretty clean implementation fixing the bug. Ready for review!

File: src/nodes/display/ViewportTextureNode.js [hunk 2]
```diff
@@ -144,9 +144,13 @@ class ViewportTextureNode extends TextureNode {
 
 	updateReference( frame ) {
 
-		const renderTarget = frame.renderer.getRenderTarget();
+		const renderer = frame.renderer;
+		const renderTarget = renderer.getRenderTarget();
+		const canvasTarget = renderer.getCanvasTarget();
 
-		this.value = this.getTextureForReference( renderTarget );
+		const reference = canvasTarget ? canvasTarget : renderTarget;
```
Comment (arodic) [inline src/nodes/display/ViewportTextureNode.js]: This line makes me think that there is underlying issue with RenderTarget/CanvasTarget API design.

@Mugen87 Can renderer have both canvasTarget and renderTarget set? If so, which case should have priority here?

Comment (sunag) [reply src/nodes/display/ViewportTextureNode.js]: `renderer.getCanvasTarget()` will never be null; the default canvas itself is defined in a CanvasTarget. In that sense, I think we should prioritize RenderTarget `renderTarget ? renderTarget : canvasTarget`, so when we have a RenderTarget, it is used as output.

Comment (arodic) [reply src/nodes/display/ViewportTextureNode.js]: @sunag, as my coworker would say "you are absolutely right!". I updated the default output to render target. The fix still appears to work as expected. I was also able to create a test case where previous version would produce an error when rendering to a RenderTarget using viewport nodes.

**Comment by arodic** [comment]
I also added tests (generated by Opus 4.5) [ViewportDepthTextureNode.tests.js‎](https://github.com/mrdoob/three.js/pull/32690/changes#diff-3579f3d5e32b2834433527cb276cb339e6aa95229c9bb4e5c1ac027f679c56c0)

Let me know if this is needed. From what I can tell, it tests for the correct edge case.

File: src/renderers/common/Renderer.js [hunk 3]
```diff
@@ -1312,7 +1311,7 @@ class Renderer {
 
 			frameBufferTarget.isPostProcessingRenderTarget = true;
 
-			this._frameBufferTarget = frameBufferTarget;
+			this._frameBufferTargets.set( canvasTarget, frameBufferTarget );
```
Comment (sunag) [inline src/renderers/common/Renderer.js:1314]: Could we add an individual dispose event for each `CanvasTarget`?

**Comment by Mugen87** [comment]
While working on some viewport/scissor related logic, I have found out this PR broke [webgpu_multiple_canvas](https://rawcdn.githack.com/mrdoob/three.js/dev/examples/webgpu_multiple_canvas.html). You see the same mesh everywhere instead of different ones like in prod:

https://threejs.org/examples/webgpu_multiple_canvas

Right before this PR, `dev` was fine.

**Comment by arodic** [comment]
@Mugen87 interesting. Ili take a look. Curious why screenshot tests didn’t pick this up?

**Comment by Mugen87** [comment]
`webgpu_multiple_canvas` is on the exception list because it's a WebGPU only demo and the CI relies on WebGL 2 (it does not support WebGPU yet).

**Comment by arodic** [comment]
@Mugen87 I drafted two PRs #33209 and #33210

Personally I'm leaning towards the latter. Lmk wdyt?

**Comment by Mugen87** [comment]
I would prefer if @sunag decides for a solution here 😇 . He knows this part of the library best.

**Comment by sunag** [comment]
@arodic Thanks for the PRs, I implemented a solution similar to yours but caching the Quads; the problem with caching the nodes after all the changes was the need to revalidate the shader, which made it a bit less efficient.
